File: src/main/java/com/oyosite/ticon/specutils/block/BlockRegistry.java
package com.oyosite.ticon.specutils.block;

import com.oyosite.ticon.specutils.*;
import com.oyosite.ticon.specutils.block.auxilary_ink_supplier.*;
import de.dafuqs.spectrum.registries.*;
import net.fabricmc.api.*;
import net.fabricmc.fabric.api.blockrenderlayer.v1.*;
import net.minecraft.client.renderer.*;
import net.minecraft.client.renderer.blockentity.*;
import net.minecraft.core.*;
import net.minecraft.core.registries.*;
import net.minecraft.resources.*;
import net.minecraft.world.item.*;
import net.minecraft.world.level.block.*;
import net.minecraft.world.level.block.entity.*;
import net.minecraft.world.level.block.state.*;

public class BlockRegistry {

	public static Block CALCITE_AUXILIARY_INK_SUPPLIER = register("calcite_auxiliary_ink_supplier", new AuxiliaryInkSupplierBlock(BlockBehaviour.Properties.of()), new Item.Properties());
	public static Block BASALT_AUXILIARY_INK_SUPPLIER = register("basalt_auxiliary_ink_supplier", new AuxiliaryInkSupplierBlock(BlockBehaviour.Properties.of()), new Item.Properties());
	public static Block MOONSTONE_GROW_LAMP = register("moonstone_grow_lamp", new com.oyosite.ticon.specutils.block.MoonstoneGrowLampBlock(BlockBehaviour.Properties.of().strength(0.6f)), new Item.Properties());

	public static CrystallarieumMaterial DRAGONBONE = new CrystallarieumMaterial.Builder()
			.createBuds(SpectrumBlocks.CRACKED_DRAGONBONE)
			.build();

	public static class BlockEntities {
		public static BlockEntityType<AuxiliaryInkSupplierBlockEntity> AUXILIARY_INK_SUPPLIER_TYPE = registerBlockEntity("auxiliary_ink_supplier", AuxiliaryInkSupplierBlockEntity::new, CALCITE_AUXILIARY_INK_SUPPLIER, BASALT_AUXILIARY_INK_SUPPLIER);
	}

	private static <T extends BlockEntity> BlockEntityType<T> registerBlockEntity(String name, BlockEntityType.BlockEntitySupplier<T> factory, Block... blocks) {
		return Registry.register(BuiltInRegistries.BLOCK_ENTITY_TYPE, SpectralUtilities.id(name), BlockEntityType.Builder.of(factory, blocks).build());
	}

	public static void register() {
		CrystallarieumMaterial.register(SpectralUtilities.id("dragonbone"), DRAGONBONE);
		NoxwoodDeco.registerAll();
	}

	private static Block register(String name, Block block, Item.Properties properties) {
		ResourceLocation id = SpectralUtilities.id(name);
		Registry.register(BuiltInRegistries.BLOCK, id, block);
		Registry.register(BuiltInRegistries.ITEM, id, new BlockItem(block, properties));
		return block;
	}

	@Environment(EnvType.CLIENT)
	public static void registerClient() {
		BlockEntityRenderers.register(BlockEntities.AUXILIARY_INK_SUPPLIER_TYPE, AuxiliaryInkSupplierBlockEntityRenderer::new);

		for (Block block : NoxwoodDeco.LAMPS.values()) {
			BlockRenderLayerMap.INSTANCE.putBlock(block, RenderType.cutout());
		}
	}

}

File: src/main/kotlin/com/oyosite/ticon/specutils/block/BlockRegistry.kt
package com.oyosite.ticon.specutils.block

import com.oyosite.ticon.specutils.SpectralUtilities.Companion.id
import com.oyosite.ticon.specutils.block.CrystallarieumMaterial.Companion.register
import com.oyosite.ticon.specutils.block.NoxwoodDeco.registerAll
import com.oyosite.ticon.specutils.block.auxilary_ink_supplier.AuxiliaryInkSupplierBlock
import com.oyosite.ticon.specutils.block.auxilary_ink_supplier.AuxiliaryInkSupplierBlockEntity
import com.oyosite.ticon.specutils.block.auxilary_ink_supplier.AuxiliaryInkSupplierBlockEntityRenderer
import de.dafuqs.spectrum.registries.SpectrumBlocks
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.fabricmc.fabric.api.blockrenderlayer.v1.BlockRenderLayerMap
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.client.renderer.blockentity.BlockEntityRenderers
import net.minecraft.core.BlockPos
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.entity.BlockEntityType.BlockEntitySupplier
import net.minecraft.world.level.block.state.BlockBehaviour
import net.minecraft.world.level.block.state.BlockState

object BlockRegistry {
    var CALCITE_AUXILIARY_INK_SUPPLIER: Block = register(
        "calcite_auxiliary_ink_supplier",
        AuxiliaryInkSupplierBlock(BlockBehaviour.Properties.of()),
        Item.Properties()
    )
    var BASALT_AUXILIARY_INK_SUPPLIER: Block = register(
        "basalt_auxiliary_ink_supplier",
        AuxiliaryInkSupplierBlock(BlockBehaviour.Properties.of()),
        Item.Properties()
    )
    var MOONSTONE_GROW_LAMP: Block = register(
        "moonstone_grow_lamp",
        MoonstoneGrowLampBlock(BlockBehaviour.Properties.of().strength(0.6f)),
        Item.Properties()
    )

    var DRAGONBONE: CrystallarieumMaterial = CrystallarieumMaterial.Builder()
        .createBuds(SpectrumBlocks.CRACKED_DRAGONBONE)
        .build()

    private fun <T : BlockEntity?> registerBlockEntity(
        name: String,
        factory: BlockEntitySupplier<T?>,
        vararg blocks: Block?
    ): BlockEntityType<T?> {
        return Registry.register<BlockEntityType<*>?, BlockEntityType<T?>>(
            BuiltInRegistries.BLOCK_ENTITY_TYPE,
            id(name),
            BlockEntityType.Builder.of<T?>(factory, *blocks).build()
        )
    }

    fun register() {
        register(id("dragonbone"), DRAGONBONE)
        registerAll()
    }

    private fun register(name: String, block: Block, properties: Item.Properties): Block {
        val id = id(name)
        Registry.register<Block?, Block?>(BuiltInRegistries.BLOCK, id, block)
        Registry.register<Item?, BlockItem?>(BuiltInRegistries.ITEM, id, BlockItem(block, properties))
        return block
    }

    @Environment(EnvType.CLIENT)
    fun registerClient() {
        BlockEntityRenderers.register<AuxiliaryInkSupplierBlockEntity?>(
            BlockEntities.AUXILIARY_INK_SUPPLIER_TYPE
        ) { context: BlockEntityRendererProvider.Context? ->
            AuxiliaryInkSupplierBlockEntityRenderer(context)
        }

        for (block in NoxwoodDeco.LAMPS.values) {
            BlockRenderLayerMap.INSTANCE.putBlock(block, RenderType.cutout())
        }

        DRAGONBONE.putBlockRenderLayerMap()
    }

    object BlockEntities {
        @JvmField
        var AUXILIARY_INK_SUPPLIER_TYPE: BlockEntityType<AuxiliaryInkSupplierBlockEntity?> =
            registerBlockEntity<AuxiliaryInkSupplierBlockEntity?>(
                "auxiliary_ink_supplier",
                { pos: BlockPos, state: BlockState? ->
                    AuxiliaryInkSupplierBlockEntity(
                        pos,
                        state
                    )
                },
                CALCITE_AUXILIARY_INK_SUPPLIER,
                BASALT_AUXILIARY_INK_SUPPLIER
            )
    }
}